.github/workflows/workflow.yml

```diff
@@ -301,14 +345,66 @@ jobs:
           init_test_run $HOME/.linkerd
           ${{ matrix.integration_test }}_integration_tests
         )
+  
+  kind_integration_gh:
+    strategy:
+      matrix:
+        integration_test: [deep, upgrade, helm, custom_domain, external_issuer]
+    needs: [docker_pull, docker_build, kind_setup]
+    name: Integration tests (${{ matrix.integration_test }})
+    runs-on: ubuntu-18.04
+    steps:
+    - name: Checkout code
+      if: github.event.pull_request.head.repo.fork
+      uses: actions/checkout@v2
+    - name: Download image archives
+      if: github.event.pull_request.head.repo.fork
+      uses: actions/download-artifact@v1
+      with:
+        name: image-archives
+    - name: Try to load cached Go modules
+      if: github.event.pull_request.head.repo.fork
+      uses: actions/cache@v1
+      with:
+        path: ~/go/pkg/mod
+        key: ${{ runner.os }}-go-${{ hashFiles('**/go.sum') }}
+        restore-keys: |
+          ${{ runner.os }}-go-
+    - name: Setup KinD (default)
+      if: github.event.pull_request.head.repo.fork && matrix.integration_test != 'custom_domain'
+      uses: engineerd/setup-kind@v0.3.0
+      with:
+        version: "v0.6.1"
+    - name: Setup KinD (custom_domain)
+      if: github.event.pull_request.head.repo.fork && matrix.integration_test == 'custom_domain'
+      uses: engineerd/setup-kind@v0.3.0
+      with:
+        config: test/testdata/custom_cluster_domain_config.yaml
+        version: "v0.6.1"
+    - name: Load image archives into KinD
+      if: github.event.pull_request.head.repo.fork
+      run: |
+        for img in controller grafana proxy web; do
+          kind load image-archive image-archives/$img.tar
+        done
+    - name: Copy out linkerd CLI
+      if: github.event.pull_request.head.repo.fork
+      run: |
+        cp image-archives/linkerd $HOME/.linkerd
+        chmod +x $HOME/.linkerd
+    - name: Run integration tests
+      if: github.event.pull_request.head.repo.fork
+      run: |
+        . bin/_test-run.sh
+        init_test_run $HOME/.linkerd
+        ${{ matrix.integration_test }}_integration_tests
 
   kind_cleanup:
-    if: always()
     strategy:
       fail-fast: false # always attempt to cleanup all clusters
       matrix:
         integration_test: [deep, upgrade, helm, custom_domain, external_issuer]
-    needs: [kind_integration]
+    needs: [kind_integration_host, kind_integration_gh]
```

Comment:
Good point. I added `kind_integration_gh` here because it still does a few of the steps (dumping env, GitHub context, and job context). However, that already happens in `validate_go_deps` and it certainly isn't needed for a matrix since there is no actual cleanup to do.

Reply:
So I tried this before and it doesn't work completely as expected. However, I may be able to make some changes.

When a conditional is applied to the job instead of the steps, the job never shows as completed like it does currently. This means that jobs that appear in `needs` contexts further down the workflow are not satisfied and stall the builds on GH builds (that didn't need an earlier job). I'll see what I can change now though since I have a better idea of the job separation.

Reply:
Okay I was able to move a lot of the conditionals to the job level.

Reply:
@kleimkuhler as per my comment above, I think you need to leave `if: always()` in `kind_cleanup` in case `kind_integration_host` fails, so we can be assured `kind_cleanup` will run regardless.

Reply:
Ah right okay thanks I'll add that back.

Reply:
Ok so I added back the `always`. The conditional for `kind_cleanup` is now `always() && (github.event_name != 'pull_request' || !github.event.pull_request.head.repo.fork)`.

This ensures that `kind_cleanup` always runs (even on cancelled workflows from a failing `kind_integration_host` job) but only on host builds.

I made sure this works in [this](https://github.com/kleimkuhler/linkerd2/actions/runs/34389215) action. I made two jobs that `always()` run; one for forked repos and one for non-forked repos. After making sure `docker_build` fails the correct job *still* ran, and the job that should run for non-forked builds did not run.